Invoke local docker

dschep opened the pull request:
This is a work in progress for implementing docker based invoke local support. 

 **Feedback greatly appreciated**

## What did you implement:
Closes #5848 by using docker with the [docker-lambda images from lambci](https://github.com/lambci/docker-lambda).
Closes #3456

The main benefits are:
 * More runtimes supported, even `provided`!
 * Layers support
 * runtime environments more closely resemble real lambda env
   * logging environment
   * paths (your handler is actually run at `/var/task`)
   * linux (as opposed to host OS)

The feature is currently implemented by default for previously unsupported runtimes, and can be enabled for python/javascript/java/ruby with the the `--docker` flag.
<!--
Briefly describe the feature if no issue exists for this PR
-->

## How did you implement it:
* local layers are resolved by traversing the `Ref` and arn based layers are downloaded and extracted into `.serverless/invokeLocal/layers/<layername>`.
* generate `.serverless/invokeLocal/Dockerfile` with a `FROM` that bases it on `lambci/lambda:${runtime}` and then `ADD`s each layer to `/opt`
* build said `Dockerfile`
* run the image with working dir mounted as `/var/task` and passing the lambda and event as the arguments

<!--
If this is a nontrivial change please briefly describe your implementation so its easy for us to understand and review your code.
-->

## How can we verify it:
* install with `npm i -g serverless/serverless#invoke-local-docker`
* create an aws project
* `sls invoke local -f hello --docker`
* add a layer and try to use it again!
<!--
Add any applicable config, commands, screenshots or other resources
to make it easy for us to verify this works. The easier you make it for us
to review a PR, the faster we can review and merge it.

Examples:
* serverless.yml - Fully functioning to easily deploy changes
* Screenshots - Showing the difference between your output and the master
* Cloud Configuration - List cloud resources and show that the correct configuration is in place (e.g. AWS CLI commands)
* Other - Anything else that comes to mind to help us evaluate
-->

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** NO
***Is it a breaking change?:*** NO

dschep commented:
I've tested with:
 * `nodejs8.10`
 * `provided`
   * bash based `aws-provided` template
   * https://github.com/dschep/sls-aws-lambci-nodejs10
 * `python3.7`
 * `go1.x`

pmuens commented on lib/plugins/aws/invokeLocal/index.js:
```diff
@@ -10,6 +10,9 @@ const chalk = require('chalk');
 const inspect = require('util').inspect;
+const download = require('download');
+const jsZip = require('jszip');
+const mkdirp = require('mkdirp');
```
Looks like some of the dependencies here need to be added to `package.json`.

In `jsZip` case it needs to be moved from `devDependencies` to `dependencies`. We could also check if `archiver` supports unzipping to save one dependency.

dschep replied:
I'll look into `archiver`, but it's not a new dep afaik... i think archiver uses it internally

dschep replied:
switched to `download` lib's built-in extraction code 🙂

pmuens commented on lib/plugins/aws/invokeLocal/index.js:
```diff
@@ -125,60 +128,75 @@ class AwsInvokeLocal {
+                    return Promise.resolve();
+                  }
+                  mkdirp.sync(
+                    path.dirname(`.serverless/invokeLocal/layers/${arnParts[6]}/${filename}`));
```
I had to replace this with:

```suggestion
                    path.join('.serverless', 'invokeLocal', 'layers', `${arnParts[6]}/${filename}`);
```

pmuens commented on lib/plugins/aws/invokeLocal/index.js:
```diff
@@ -125,60 +128,75 @@ class AwsInvokeLocal {
+        dockerfile += `\nADD --chown=sbx_user1051:495 ${layerPath} /opt`;
+        // fs.copySync(layerPath, `.serverless/invokeLocal/${layerPath}`);
+      }
+      mkdirp.sync(path.dirname('.serverless/invokeLocal'));
```
I had to replace this with:

```suggestion
      mkdirp.sync(path.join('.serverless' , 'invokeLocal'));
```

pmuens commented on lib/plugins/aws/invokeLocal/index.js:
```diff
@@ -125,60 +128,75 @@ class AwsInvokeLocal {
+      fs.writeFileSync('.serverless/invokeLocal/Dockerfile', dockerfile);
+      this.serverless.cli.consoleLog('Building docker image...');
+      const docker = spawn('docker', ['build', '-t', imageName,
+        `${this.serverless.config.servicePath}`, '-f', '.serverless/invokeLocal/Dockerfile']);
```
Same question about a potential `path.join` usage here.

pmuens commented on lib/plugins/aws/invokeLocal/index.js:
```diff
@@ -125,60 +128,75 @@ class AwsInvokeLocal {
+      }
+      mkdirp.sync(path.dirname('.serverless/invokeLocal'));
+      fs.writeFileSync('.serverless/invokeLocal/Dockerfile', dockerfile);
+      this.serverless.cli.consoleLog('Building docker image...');
```
Do we keep this log later on or do we turn this into `this.serverless.cli.log` for a better formatting?

We could also break this down into separate steps a like "Downloading Image", "Buliding container" or something like that because it could take quite some time for larger images and slow internet connections.

dschep replied:
Ah, oops. Just using the wrong function. As for splitting it, good idea 👍

dschep commented:
Switched to `path.join`, the right logging func, and split download into its own step.

Also thinking layer content should be cached between invokations (SAM does this too). Should that live in a global cache/tmp dir? or in .serverless? Do we have any caching in the framework currently?

pmuens commented on lib/plugins/aws/invokeLocal/index.js:
```diff
@@ -177,8 +189,145 @@ class AwsInvokeLocal {
+      || this.serverless.service.provider.runtime
+      || 'nodejs4.3';
+
+    this.serverless.cli.log('Downloading base docker image...');
```
Probably personal preference, but I'd vote for `Docker` vs. `docker` since it's a company / product name.

pmuens commented on lib/plugins/aws/invokeLocal/index.js:
```diff
@@ -177,8 +189,145 @@ class AwsInvokeLocal {
+          if (fs.existsSync(layerContentsPath)) {
+            return layerContentsPath;
+          }
+          let downloadPromise = Promise.resolve();
```
Same Promise-usage comment applies here.

pmuens commented on lib/plugins/aws/invokeLocal/index.test.js:
```diff
@@ -468,10 +472,23 @@ describe('AwsInvokeLocal', () => {
+      awsInvokeLocal.options.functionObj.handler = 'handler.foobar';
+      return awsInvokeLocal.invokeLocal().then(() => {
+        expect(invokeLocalDockerStub.calledOnce).to.be.equal(true);
+        expect(invokeLocalDockerStub.calledWithExactly()).to.be.equal(true);
```
What's the difference between this and `calledOnce`?

dschep replied:
`calledOnce` ensures that the function wasn't called more than one time. `calledWithExactly` ensures that it was called _at least once with those exact arguments_, IE, it could be called any number of additional times with other arguments.

pmuens replied:
Makes sense...

pmuens commented on lib/plugins/aws/invokeLocal/index.test.js:
```diff
@@ -468,10 +472,23 @@ describe('AwsInvokeLocal', () => {
+      awsInvokeLocal.options.docker = true;
+      return awsInvokeLocal.invokeLocal().then(() => {
+        expect(invokeLocalDockerStub.calledOnce).to.be.equal(true);
+        expect(invokeLocalDockerStub.calledWithExactly()).to.be.equal(true);
```
Same question about `calledOnce()` vs. `calledWithExactly()` here.

pmuens commented on lib/plugins/aws/invokeLocal/index.test.js:
```diff
@@ -1097,4 +1114,85 @@ describe('AwsInvokeLocal', () => {
+      delete require.cache[require.resolve('child_process')];
+    });
+
+    it('watwat', () =>
```
😅

pmuens commented on lib/plugins/aws/invokeLocal/index.js:
```diff
@@ -177,8 +189,145 @@ class AwsInvokeLocal {
+    }
+    return fs.readFileAsync(artifact)
+      .then(jszip.loadAsync)
+      .then(zip => Promise.all(
```
Same Promise-usage comment applies here.

pmuens commented on lib/plugins/aws/invokeLocal/index.js:
```diff
@@ -177,8 +189,145 @@ class AwsInvokeLocal {
+        Object.keys(zip.files)
+          .map(filename => zip.files[filename].async('nodebuffer').then(fileData => {
+            if (filename.endsWith(path.sep)) {
+              return Promise.resolve();
```
Same Promise-usage comment applies here.

pmuens commented on lib/plugins/aws/invokeLocal/index.js:
```diff
@@ -177,8 +189,145 @@ class AwsInvokeLocal {
+  invokeLocalDocker() {
+    const handler = this.options.functionObj.handler;
+
+    return Promise.all([
```
Same Promise-usage comment applies here.

pmuens commented on lib/plugins/aws/invokeLocal/index.js:
```diff
@@ -177,8 +189,145 @@ class AwsInvokeLocal {
+      (value, key) => this.provider.naming.getLambdaLayerLogicalId(key)
+    );
+
+    return Promise.all(
```
Best would be to stick to one Promise implementation (`BbPromise` in this case).

Reason being that we once hunted down a race condition in the Serverless Variables implementation for days after figuring out that we used native Promises and Bluebird Promises in the same codebase and one implementation (probably Bluebird) was slightly faster.

pmuens commented on lib/plugins/aws/invokeLocal/index.js:
```diff
@@ -177,8 +189,145 @@ class AwsInvokeLocal {
+      mkdirp.sync(path.join('.serverless', 'invokeLocal'));
+      const dockerfilePath = path.join('.serverless', 'invokeLocal', 'Dockerfile');
+      fs.writeFileSync(dockerfilePath, dockerfile);
+      this.serverless.cli.log('Building docker image...');
```
Probably personal preference, but I'd vote for `Docker` vs. `docker` since it's a company / product name.

dschep replied:
Fair enough. When I think [Dd]ocker, I think `docker` bc it's `/usr/bin/docker` not `/usr/bin/Docker` 😜

pmuens replied:
Make sense! 😊

sheerun commented:
I get following error:

Serverless plugin "/usr/local/lib/node_modules/serverless/lib/plugins/aws/invokeLocal/index.js" initialization errored: Cannot find module 'jszip'

triptec commented:
@dschep seems jszip is listed under devDependencies so that might need to be moved to devependencies. Thanks for the work by the way, looking forward to having this merged.

pmuens commented on lib/plugins/aws/invokeLocal/index.js:
```diff
@@ -177,8 +189,145 @@ class AwsInvokeLocal {
+  }
+
+  buildDockerImage(layerPaths) {
+    const runtime = this.options.functionObj.runtime
```
This piece of code is used at least 4 times in this file alone. Maybe we can use a function like `getRuntime(this)` to DRY that out a little bit...

endeepak commented on lib/plugins/aws/invokeLocal/index.js:
```diff
@@ -177,8 +192,151 @@ class AwsInvokeLocal {
+              'Lambda', 'getLayerVersion', { LayerName: layerArn, VersionNumber: layerVersion })
+              .then(layerInfo => download(
+                layerInfo.Content.Location,
+                layerContentsPath,
```
@dschep I think this should be downloading to `layerContentsCachePath` instead of `layerContentsPath`. Currently this code creates empty directory in `layerContentsCachePath` in first run and subsequent runs will have empty directory copied to `layerContentsPath`. Due to this handler fails to import layer in subsequent runs

endeepak replied:
I've raised an issue for this -> https://github.com/serverless/serverless/issues/5990

dschep replied:
Good catch
